### PR TITLE
RF: Implement new API (add)

### PR DESCRIPTION
**Pull request by bpoldrack**

### CODE REVIEW

**Comment by hanke** [comment]
Note: this command should query a datasets configuration to decide how to add files. For example, I might want to configure that any files in `code/` are always added to git directly.

File: datalad/distribution/add.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+                        ds.repo.add_url_to_file(f, u,
+                                                options=annex_add_opts,  # TODO: see above
+                                                git_options=git_opts,
+                                                annex_options=annex_opts,
+                                                batch=True))
+
+        return return_values
+
+    @staticmethod
+    def result_renderer_cmdline(res):
```
Comment (mih) [inline datalad/distribution/add.py]: API has changed: see https://github.com/datalad/datalad/pull/740

File: datalad/distribution/add.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            doc="""flag whether to add data directly to Git, instead of
+            tracking data identity only.  Usually this is not desired,
+            as it inflates dataset sizes and impacts flexibility of data
+            transport"""),
+        recursive=recursion_flag,
+        recursion_limit=recursion_limit,
+        git_opts=git_opts,
+        annex_opts=annex_opts,
+        annex_add_opts=annex_add_opts)
+
```
Comment (mih) [inline datalad/distribution/add.py]: @bpoldrack @yarikoptic I guess we should decide how/if we want to support saving changing in each interface. Aggregate does it, this one doesn't. I see boths pros and cons for both. Adding this option makes things more convoluted, but when ON by default we could tailor tthe resulting commits to be more meaningful -- compared to a `save -a -r` after some time. Sure thing there should always be a way to _not_ auto-save things.

**Comment by mih** [comment]
Observation:

````
% touch one two three
% datalad add one two
Added one
Added two
% datalad add one two
Nothing was added

# expected so far, but
% datalad add one two --to-git
Added one
Added two
% datalad add one two --to-git
Added one
Added two
````

**Comment by mih** [comment]
Recursive does not work as expected (continued scenario from above):

````
% datalad create deep --add-to-super
Created dataset at /tmp/some/deep.
% datalad save
% datalad add -r .
Added three
% datalad ls -r
.     [annex]  master  - 2016-09-01/09:14:03  X
deep  [annex]  master  - 2016-09-01/09:13:52  X
% git -C deep status
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	some

nothing added to commit but untracked files present (use "git add" to track)
````

I guess it doesn't realize that `.` matches (or should match IHMO) the submodule too.

**Comment by mih** [comment]
Otherwise this is looking good, I think. I'd appreciate quick fixes and a swift merge (that I'd myself ;-)

**Comment by mih** [comment]
See https://github.com/bpoldrack/datalad/pull/30 for some trivial fixes.

File: datalad/distribution/add.py (hunk 2 above)
Comment (bpoldrack) [inline datalad/distribution/add.py]: I vote for always providing an autosave option. Makes sense to me.

**Comment by bpoldrack** [comment]
Regarding recursion: Currently unclear to me, whether this is what we want. Need to talk, I guess. Note, that `datalad add * -r` works as expected.

**Comment by bpoldrack** [comment]
Regarding your observation of reporting success twice in case of  `--to-git`:
This is due to the way I currently generate a return value for `git add`. Won't fix in this PR. It's a more general thing, that is to solve within `GitRepo`.

**Comment by mih** [comment]
FTR: `--recursive` means "recurse into subdatasets. Otherwise recursion into subdirectories is always implied (just like git annex does it).

IMHO: any argument needs to be resolved against the root of the respective dataset (that had its `add()` called). If any of the given paths is a directory, any subdataset below this directory needs to to receive `'.'` as the sole path argument to its `add()` call.

**Comment by bpoldrack** [comment]
Agree. That's exactly, what I'm currently implementing.

**Comment by mih** [comment]
Tests fail, but otherwise looks good.

**Comment by bpoldrack** [comment]
Strange. Apparently, Travis ran on a master, that I didn't merge yet, while the buildbots used an older one and passed.

**Comment by mih** [comment]
OK, let's merge this now and unify the behavior re save in one go, once the required utilities are in.

**Comment by yarikoptic** [comment]
FYI merge caused major havoc conflict in #746 which I hopefully resolve correctly now

**Comment by yarikoptic** [comment]
actually I am wrong -- this one alone couldn't do it because of only a tiny change (force arg)... heh -- smth else then

**Pull request merged**